tests/import_specifications/test_file_parser.py

```diff
@@ -3,37 +3,110 @@
 from frozendict import frozendict
 from pytest import raises
 from tests.test_utils import assert_exception_correct
+from pathlib import Path
 
 from staging_service.import_specifications.file_parser import (
     ErrorType,
+    FileTypeResolution,
     SpecificationSource,
     Error,
+    ParseResult,
     ParseResults,
 )
 
+def spcsrc(path, tab=None):
+    return SpecificationSource(Path(path), tab)
+
+
+def test_SpecificationSource_init_success():
+    # minimal
+    ss = SpecificationSource(Path("foo"))
+
+    assert ss.file == Path("foo")
+    assert ss.tab is None
+
+    # all
+    ss = SpecificationSource(Path("bar"), "tabbytab")
+
+    assert ss.file == Path("bar")
+    assert ss.tab == "tabbytab"
+
+
+def test_SpecificationSource_init_fail():
+    # could inline this, but might as well follow the same pattern as all the other tests
+    specificationSource_init_fail(None, ValueError("file is required"))
+
+
+def specificationSource_init_fail(file_, expected):
+    with raises(Exception) as got:
+        SpecificationSource(file_)
+    assert_exception_correct(got.value, expected)
+
+
+def test_FileTypeResolution_init_w_parser_success():
+    p = lambda path: ParseResults(errors=(Error(ErrorType.OTHER, "foo"),))
+    ftr = FileTypeResolution(p)
+
+    assert ftr.parser is p  # Here only identity equality makes sense
+    assert ftr.unsupported_type is None
+
+
+def test_FileTypeResolution_init_w_unsupported_type_success():
+    ftr = FileTypeResolution(unsupported_type="sys")
+
+    assert ftr.parser is None
+    assert ftr.unsupported_type == "sys"
+
+
+def test_FileTypeResolution_init_fail():
+    err = "Exectly one of parser or unsupported_type must be supplied"
+    pr = ParseResults(errors=(Error(ErrorType.OTHER, "foo"),))
+    fileTypeResolution_init_fail(None, None, ValueError(err))
+    fileTypeResolution_init_fail(lambda path: pr, "mp-2", ValueError(err))
+
+
+def fileTypeResolution_init_fail(type_, utype, expected):
```

Comment:
Can we add a docstring here

Reply:
I generally don't add docstrings in test code unless there's something really confusing going on that needs comments. This function seems to be an outlier in that you think it needs docstrings where other similar functions don't - I suspect this is because of the crappy variable names I used. I've provided better variable names in hopes that that takes care of the issue, but if that's not enough let me know.